packages/aws-cdk-lib/aws-cloudfront-origins/lib/s3-bucket-origin.ts

```diff
@@ -151,32 +157,25 @@ class S3BucketOriginWithOAC extends S3BucketOrigin {
     };
   }
 
-  private getBucketPolicyActions(accessLevels: Set<cloudfront.AccessLevel>): string[] {
-    let actions: string[] = [];
-    for (const accessLevel of accessLevels) {
-      actions = actions.concat(BUCKET_ACTIONS[accessLevel]);
-    }
-    return actions;
+  private getBucketPolicyActions(accessLevels: Set<cloudfront.AccessLevel>): BucketPolicyAction[] {
+    return [...accessLevels].flatMap((accessLevel) => BUCKET_ACTIONS[accessLevel] ?? []);
   }
 
   private getKeyPolicyActions(accessLevels: Set<cloudfront.AccessLevel>): string[] {
-    let actions: string[] = [];
-    for (const accessLevel of accessLevels) {
-      // Filter out DELETE since delete permissions are not applicable to KMS key actions
-      if (accessLevel !== AccessLevel.DELETE) {
-        actions = actions.concat(KEY_ACTIONS[accessLevel]);
-      }
-    }
-    return actions;
+    return [...accessLevels].flatMap((accessLevel) => KEY_ACTIONS[accessLevel] ?? []);
   }
 
-  private grantDistributionAccessToBucket(distributionId: string, actions: string[]): iam.AddToResourcePolicyResult {
+  private grantDistributionAccessToBucket(distributionId: string, actions: BucketPolicyAction[]): iam.AddToResourcePolicyResult {
+    const resources = [this.bucket.arnForObjects('*')];
+    if (actions.some(({ bucketArn }) => bucketArn)) {
+      resources.push(this.bucket.bucketArn);
+    }
```

Comment:
Can you explain this part please?

Reply:
s3:ListBucket action requires bucket ARN in `resources` (instead of object ARN required by s3:GetObject etc.)
see https://docs.aws.amazon.com/service-authorization/latest/reference/list_amazons3.html

Reply:
When READ and LIST are specified, `actions` will be `[{ action: 's3:GetObject' }, { action: 's3:ListBucket', bucketArn: true }]`.
The if condition returns whether `actions` has one or more elements with truthy `bucketArn` prop, so the result will be true. Then the the bucketArn will be pushed to `resources`.

It is the reason why the object arn (`arnForObjects('*')`) is always in resources that the LIST permission should be used with READ in most cases.

Reply:
updated names more descriptive.